### DISCUSSION TITLE
classes/class_gradient

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#242] [User-contributed notes]
# classes/class_gradient

Inherits: Resource< RefCounted< Object A color transition. Description: This resource describes a color transition by defining a set of colored points and how to interpolate between them. See also ...

https://docs.godotengine.org/en/stable/classes/class_gradient.html

**Comment by grdleo** [comment]
When creating a new gradient, it has already two existing points.
This is particularly annoying when one wants to create a full custom gradient.
Using `Gradient.new()` should create an empty gradient imo, or at least the user should have the opportunity to create an empty gradient from the API.

It should not have to do this
```gdscript
var g = Gradient.new()
g.remove_point(0)
g.remove_point(0)
...
```

in my script.

**Reply by Calinou** [reply]
Changing this behavior would break compatibility in existing projects, so it can't be done now. You can still open a [proposal](https://github.com/godotengine/godot-proposals) for this, but it can't be implemented before Godot 5.0 unless it's done in a backwards-compatible manner, e.g. by adding a `Gradient.create_empty()` static method.

**Reply by kraftwer1** [reply]
Thanks a lot. I just wanted to point out that I had to remove those default colors *after* I have added my own colors, in order to prevent the error that `Condition "points.size() <= 1" is true.`:

```
var g = Gradient.new()
g.add_point(…)
g.add_point(…)

# Remove default colors
g.remove_point(0)
g.remove_point(0)
...
```

**Comment by ratman-studio** [comment]
in my opinion it should be something like `Gradient.FROM_COLORS(colors:Array[Color])`
here is my code to create gradient with equal distribution from `Array[Color]`
```gdscript
func create_gradient(colors: Array[Color]) -> Gradient:
	var gradient    = Gradient.new() 	# creates black to white gradient with two points

	# check colors array size
	var amount: int = colors.size()
	if amount < 1:
		printerr("gradient cant have less than one color")
		return gradient # return default gradient

	# remove second color if only one color is provided
	if amount == 1: gradient.remove_point(1)

	# set or add new colors in equal intervals from 0 to 1
	for i in range(amount):
		var position = lerp(0, 1, i/(amount-1.0))

		if gradient.get_point_count() <= i:
			gradient.add_point(position, colors[i])

		else:
			gradient.set_color(i, colors[i])
			gradient.set_offset(i, position)

	return gradient
```

**Comment by NicholasPetela** [comment]
sample function should probably mention that values should be from 0.0 to 1.0, inclusive.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/107976
